### PR TITLE
Add garbage collection

### PR DESCRIPTION
**Pull request by Nightcro**
Please make sure these boxes are checked, before submitting a new PR.

- [x] You ran `rustfmt` on the code base before submitting (on latest nightly with rustfmt support)
- [x] You reference which issue is being closed in the PR text (if applicable)

## Issues

Added garbage collection.
Added 1 test for checking.

At the moment the implementation is barebone. The user is not aware if the gc succeeded or not. The replicas do not answer back to the leader.

### CODE REVIEW

File: src/storage.rs [hunk 1]
```diff
@@ -96,6 +96,9 @@ where
     /// Returns true if the log contains a StopSign or a StopSign already has been decided.
     /// Note that the log could have a StopSign that later gets overwritten, and thus this function might first return true and later false.
     fn stopped(&self) -> bool;
+
+    /// Starts the garbage collection process to free storage.
```
Comment (haraldng) [inline src/storage.rs:100]: This function does not start the garbage collection process (which is done in the `OmniPaxos` struct), it is rather just removing the first `idx` elements of the log.

```suggestion
    /// Removes elements up to the given [`idx`] from storage.
```

File: src/storage.rs [hunk 2]
```diff
@@ -123,6 +126,12 @@ where
 
     /// Returns the round that has been promised.
     fn get_promise(&self) -> R;
+
+    /// Sets the garbage collector index.
```
Comment (haraldng) [inline src/storage.rs:130]:
```suggestion
    /// Sets the garbage collected index.
```

File: src/storage.rs [hunk 3]
```diff
@@ -123,6 +126,12 @@ where
...
+    fn set_gc_idx(&mut self, index: u64);
+
+    /// Returns the garbage collector index.
```
Comment (haraldng) [inline src/storage.rs:133]:
```suggestion
    /// Returns the garbage collected index.
```

File: tests/config/test.conf [hunk 4]
```diff
@@ -22,4 +22,14 @@ proposal_test {
     num_nodes: 6
```
Comment (haraldng) [inline tests/config/test.conf:22]: Why is this 6? Use uneven numbers

File: tests/config/test.conf [hunk 5]
```diff
@@ -22,4 +22,14 @@ proposal_test {
...
     ble_hb_delay: 5
     increment_delay: 2
-}
\ No newline at end of file
+}
+
+gc_test {
+    wait_timeout: 2 s
+    num_threads: 8
+    num_nodes: 11
```
Comment (haraldng) [inline tests/config/test.conf:30]: I don't think 11 nodes are needed, consensus clusters are usually at most 7 nodes.

File: tests/util.rs [hunk 6]
```diff
@@ -393,6 +393,18 @@ pub mod omnireplica {
...
+        }
+
+        pub fn garbage_collect_by_index(&mut self, index: u64) {
+            self.paxos.garbage_collect_by_index(index)
+        }
+
+        pub fn get_garbage_collector_index(&self) -> u64 {
+            self.paxos.get_gc_from_storage()
+        }
+
```
Comment (haraldng) [inline tests/util.rs:407]: Do we really need these functions if their body is one line anyway? :D

File: src/paxos.rs [hunk 7]
```diff
@@ -189,6 +191,72 @@ where
...
     }
 
+    /// Initiates the garbage collection process using the provided index.
+    /// # Arguments
+    /// * `index` - Deletes all entries up to [`index`]
+    pub fn garbage_collect_by_index(&mut self, index: u64) {
+        self.gc_propose(Some(index));
+    }
+
+    /// Initiates the garbage collection process using the minimum last accepted sequence as the [`index`].
```
Comment (haraldng) [inline src/paxos.rs:201]:
```suggestion
    /// Initiates the garbage collection process using the minimum index accepted by **ALL** servers as the [`index`].
```

File: src/paxos.rs [hunk 8]
```diff
@@ -189,6 +191,72 @@ where
...
+    pub fn garbage_collect(&mut self) {
+        self.gc_propose(None);
+    }
+
+    /// Return garbage collection index from storage.
+    pub fn get_gc_from_storage(&self) -> u64 {
+        self.storage.get_gc_idx()
+    }
+
+    fn gc_propose(&mut self, index: Option<u64>) {
```
Comment (haraldng) [inline src/paxos.rs:211]: This function appear unnecessary... just put its body into `garbage_collect`

File: src/paxos.rs [hunk 9]
```diff
@@ -443,6 +522,19 @@ where
         }
     }
 
+    fn forward_gc_proposal(&mut self, index: Option<u64>) {
+        if self.leader > 0 && self.leader != self.pid {
+            trace!(
+                self.logger,
+                "Forwarding gc proposal to Leader {}",
```
Comment (haraldng) [inline src/paxos.rs:529]: "gc request" and the index of it

File: src/paxos.rs [hunk 10]
```diff
@@ -189,6 +191,72 @@ where
...
+            );
+            return;
+        }
+
+        trace!(self.logger, "Garbage Collection index: {:?}", index);
+
+        self.cached_gc_index = index.unwrap();
+
+        self.storage.set_gc_idx(index.unwrap());
+        self.storage.garbage_collect(index.unwrap());
```
Comment (haraldng) [inline src/paxos.rs:257]: we **must** call `self.storage.garbage_collect()` first. We don't want to first set the index and then crash and when we recover we will have the updated gc_index but the entries are still in the log.

File: src/storage.rs [hunk 11]
```diff
@@ -308,6 +317,28 @@ where
...
+                s.garbage_collect(idx);
+                self.paxos_state.set_gc_idx(idx);
+            }
+            PaxosSequence::Stopped(_) => {} // todo what to do when paxos is stopped?
+            _ => panic!("Got unexpected intermediate PaxosSequence::None in stopped()"),
+        }
+    }
+
+    /// Sets the garbage collector index in storage.
+    pub fn set_gc_idx(&mut self, index: u64) {
```
Comment (haraldng) [inline src/storage.rs:334]: Remove this function. The storage should always perform the gc and then set the gc_idx internally. A user should not be able to only set the gc_idx without actually garbage collecting.

**Review by haraldng — CHANGES_REQUESTED** [review]

File: tests/util.rs (hunk 6 above)
Comment (Nightcro) [reply tests/util.rs:407]: I can delete the first one, but the other two I need for the tests

File: src/paxos.rs [hunk 12]
```diff
@@ -189,6 +191,83 @@ where
...
+                decided_len
+            );
+            return;
+        }
+
+        trace!(self.logger, "Garbage Collection index: {:?}", index);
+
+        self.cached_gc_index = index;
+
+        self.storage.garbage_collect(index);
```
Comment (haraldng) [inline src/paxos.rs:268]: Please put `self.storage.garbage_collect(index)` before `self.cached_gc_index = index`

File: src/storage.rs [hunk 13]
```diff
@@ -308,6 +317,23 @@ where
...
     }
+
+    /// Removes elements up to the given [`idx`] from storage.
+    pub fn garbage_collect(&mut self, idx: u64) {
+        match self.sequence {
+            PaxosSequence::Active(ref mut s) => {
+                s.garbage_collect(idx - self.paxos_state.get_gc_idx());
+                self.paxos_state.set_gc_idx(idx);
+            }
+            PaxosSequence::Stopped(_) => {} // todo what to do when paxos is stopped?
```
Comment (haraldng) [inline src/storage.rs:328]: If we are stopped, we can carbage collect too. Here, I think we can actually use the OR `|` in pattern matching :D 

```suggestion
            PaxosSequence::Active(ref mut s) | PaxosSequence::Stopped(_) => {
                s.garbage_collect(idx - self.paxos_state.get_gc_idx());
                self.paxos_state.set_gc_idx(idx);
            }
```

File: src/storage.rs [hunk 14]
```diff
@@ -393,6 +423,8 @@ pub mod memory_storage {
         acc_round: R,
         /// Length of the decided sequence.
         ld: u64,
+        /// Garbage collector index.
```
Comment (haraldng) [inline src/storage.rs:426]:
```suggestion
        /// Garbage collected index.
```

**Review by haraldng — CHANGES_REQUESTED** [review]
Nice! Just one more thing apart from the comments: Can we extend the gc_test so we perform multiple garbage collections, one after the other.

**Review by haraldng — APPROVED** [review]

**Pull request merged**